pkg/controller/deorbit/deorbiter.go

```diff
@@ -0,0 +1,181 @@
+	PollInterval = 15 * time.Second
+)
+
+type Deorbiter interface {
+	DeletePersistentVolumeClaims() ([]core_v1.PersistentVolumeClaim, error)
+	DeleteServices() ([]core_v1.Service, error)
+	WaitForPersistentVolumeCleanup() error
+	WaitForServiceCleanup() error
+	SelfDestruct(SelfDestructReason) error
+}
+
+type ConcreteDeorbiter struct {
+	Kluster *v1.Kluster
+	Stop    <-chan struct{}
+	Client  kubernetes.Interface
+	Logger  log.Logger
+}
+
+func NewDeorbiter(kluster *v1.Kluster, stopCh <-chan struct{}, clients config.Clients, recorder record.EventRecorder, logger log.Logger) (Deorbiter, error) {
+	client, err := clients.Satellites.ClientFor(kluster)
+	if err != nil {
+		return nil, err
+	}
+
+	var deorbiter Deorbiter
+	deorbiter = &ConcreteDeorbiter{kluster, stopCh, client, logger}
+	deorbiter = &LoggingDeorbiter{deorbiter, logger}
+	deorbiter = &EventingDeorbiter{deorbiter, kluster, recorder}
+	deorbiter = &InstrumentingDeorbiter{
+		deorbiter,
+		metrics.DeorbitOperationsLatency,
+		metrics.DeorbitOperationsTotal,
+		metrics.DeorbitSuccessfulOperationsTotal,
+		metrics.DeorbitFailedOperationsTotal,
+	}
+
+	return deorbiter, nil
+}
+
+func (d *ConcreteDeorbiter) DeletePersistentVolumeClaims() (deleted []core_v1.PersistentVolumeClaim, err error) {
+	deleted = []core_v1.PersistentVolumeClaim{}
+
+	pvcs, err := d.Client.Core().PersistentVolumeClaims(meta_v1.NamespaceAll).List(meta_v1.ListOptions{})
+	if err != nil {
+		return deleted, err
+	}
+
+	for _, pvc := range pvcs.Items {
+		// Only PVCs backed by Cinder are interessting for us. Extra secure nil check
+		// to not accidentally dereference nil
+		if pvc.Spec.StorageClassName == nil || *pvc.Spec.StorageClassName != "cinder-default" {
+			continue
+		}
+		deleted = append(deleted, pvc)
+
+		err = d.Client.Core().PersistentVolumeClaims(pvc.Namespace).Delete(pvc.Name, &meta_v1.DeleteOptions{})
+		if err != nil {
+			return deleted, err
+		}
+	}
+
+	return deleted, err
+}
+
+func (d *ConcreteDeorbiter) DeleteServices() (deleted []core_v1.Service, err error) {
+	deleted = []core_v1.Service{}
+
+	services, err := d.Client.Core().Services(meta_v1.NamespaceAll).List(meta_v1.ListOptions{})
+	if err != nil {
+		return deleted, err
+	}
+
+	for _, service := range services.Items {
+		if service.Spec.Type != "LoadBalancer" {
+			continue
+		}
+		deleted = append(deleted, service)
+
+		err = d.Client.Core().Services(service.Namespace).Delete(service.Name, &meta_v1.DeleteOptions{})
+		if err != nil {
+			return deleted, err
+		}
+	}
+
+	return deleted, err
+}
+
+func (d *ConcreteDeorbiter) WaitForPersistentVolumeCleanup() (err error) {
+	done, err := d.isPersistentVolumesCleanupFinished()
+	if err != nil {
+		return err
+	}
+
+	if done {
+		return nil
+	}
+
+	return wait.PollUntil(PollInterval, d.isPersistentVolumesCleanupFinished, d.Stop)
+}
+
+func (d *ConcreteDeorbiter) WaitForServiceCleanup() (err error) {
+	done, err := d.isPersistentVolumesCleanupFinished()
```

Comment:
Why is `isPersistentVolumesCleanupFinished` called in `WaitForServiceCleanup`?

Reply:
Yes, copy-paste foo. I'm for sure going to test it all again before merge.